File: cvc-plugin/src/lsp/protocol.ts
/**
 * CVC LSP Protocol Types
 *
 * These types mirror the Rust protocol types in cvc-lsp/src/protocol.rs
 * and define the custom LSP notifications used by CVC.
 */

/**
 * Author of an interaction - matches cvc-core's Author enum
 */
export type Author = "human" | "agent" | "system" | "external";

/** Read-only policy state. It contains no interaction or remote identity data. */
export interface PrivacyStatus {
  captureAcknowledged: boolean;
  captureNoticeVersion: number;
  passiveCaptureAllowed: boolean;
  privateByDefault: boolean;
  privateDefaultStatement: string;
  sharingSummary: string;
  autoPushEnabled: boolean;
}

/**
 * Parameters for $/cvc/session/start notification
 * Signals the beginning of a logical task (e.g., opening a chat window)
 */
export interface SessionStartParams {
  /** Title/description of the session */
  title: string;
  /** Unix timestamp (optional, server will use current time if not provided) */
  timestamp?: number;
}

/**
 * Parameters for $/cvc/turn/start notification
 * Sent when user initiates a prompt/request
 */
export interface TurnStartParams {
  /** Correlation ID to match with turn/end */
  id: string;
  /** The user's prompt/question */
  prompt: string;
  /** Source of the prompt */
  author: Author;
  /** Paths of files in context (optional) */
  contextFiles?: string[];
}

/**
 * Parameters for $/cvc/turn/end notification
 * Sent when the model finishes responding
 */
export interface TurnEndParams {
  /** Correlation ID matching the turn/start */
  id: string;
  /** The model's response text */
  response?: string;
  /** Chain of thought / reasoning (if available) */
  chainOfThought?: string;
  /** Model name/identifier */
  model?: string;
  /** Raw response parts for robust parsing on server */
  rawResponse?: unknown[];
}

/**
 * Parameters for $/cvc/turn/batch notification
 * Sends a complete set of segmented interactions for a single VS Code chat request.
 * Used by the Chat Session Watcher for retroactive parsing.
 */
export interface TurnBatchParams {
  /** The VS Code requestId - used for deduplication */
  sourceRequestId: string;
  /** The session/conversation ID */
  sessionId: string;
  /** Model name/identifier */
  model?: string;
  /** Ordered list of interaction segments */
  interactions: InteractionSegment[];
}

/**
 * A single segment of a chat interaction.
 * The first segment is typically author: "human" with the prompt.
 * Subsequent segments are author: "agent" with chain of thought + response.
 */
export interface InteractionSegment {
  /** Author of this segment */
  author: Author;
  /** User prompt (only on human segments) */
  userPrompt?: string;
  /** Chain of thought / thinking block (only on agent segments) */
  chainOfThought?: string;
  /** Visible response text */
  response?: string;
  /** Context files (typically only on human segment) */
  contextFiles?: string[];
}

/**
 * Parameters for $/cvc/link/commit notification
 * Associates interactions with a Git commit
 */
export interface LinkCommitParams {
  /** The Git commit SHA */
  commitSha: string;
  /** IDs of interactions to link to this commit */
  interactionIds: string[];
}

/**
 * Parameters for cvc/timeline/get request (future feature)
 * Requests the cognitive timeline for display
 */
export interface TimelineGetParams {
  /** Maximum number of items to return */
  maxItems?: number;
  /** Include unbound/floating thoughts */
  includeUnbound?: boolean;
  /** The SHA of the current HEAD commit to filter by */
  headSha?: string;
}

/**
 * Response structure for cvc/timeline/get (future feature)
 */
export interface TimelineGetResponse {
  /** Floating/pending thoughts not yet linked to commits */
  pending: InteractionSummary[];
  /** Commits with their linked thoughts */
  commits: CommitWithThoughts[];
}

/**
 * Summary of an interaction for timeline display
 */
export interface InteractionSummary {
  /** Unique interaction ID */
  id: string;
  /** Truncated/preview of the prompt */
  promptPreview: string;
  /** Unix timestamp */
  timestamp: number;
  /** Author type */
  author: Author;
  /** Flags indicating which content types are present */
  hasPrompt: boolean;
  hasCot: boolean;
  hasResponse: boolean;
}

/**
 * A commit with its linked thoughts
 */
export interface CommitWithThoughts {
  /** Git commit SHA */
  sha: string;
  /** Commit message (first line) */
  message: string;
  /** Unix timestamp of commit */
  timestamp: number;
  /** Thoughts linked to this commit */
  thoughts: InteractionSummary[];
}

/**
 * Parameters for cvc/interaction/get request
 * Fetches full details of a specific interaction
 */
export interface InteractionGetParams {
  /** The interaction ID to fetch */
  id: string;
}

/**
 * Full interaction details returned by cvc/interaction/get
 */
export interface InteractionDetail {
  /** Unique interaction ID */
  id: string;
  /** Conversation this interaction belongs to */
  conversationId: string;
  /** Parent interaction ID (if any) */
  parentId?: string;
  /** Unix timestamp */
  timestamp: number;
  /** Author of the prompt */
  author: Author;
  /** Full user prompt */
  userPrompt: string;
  /** Model name used */
  modelName?: string;
  /** Chain of thought / reasoning */
  chainOfThought?: string;
  /** Model's response */
  modelResponse?: string;
  /** Context files attached to this interaction */
  contextFiles?: ContextFileInfo[];
  /** Tool executions during this interaction */
  toolExecutions?: ToolExecutionInfo[];
  /** Linked commit SHA (if bound) */
  linkedCommit?: string;
}

/**
 * Information about a context file
 */
export interface ContextFileInfo {
  /** File path */
  path: string;
  /** Git blob SHA (if file was clean) */
  blobSha?: string;
  /** Line range (if partial) */
  startLine?: number;
  endLine?: number;
}

/**
 * Information about a tool execution
 */
export interface ToolExecutionInfo {
  /** Tool name */
  name: string;
  /** Tool protocol (mcp, native, etc.) */
  protocol: string;
  /** Arguments passed to the tool */
  arguments?: string;
  /** Execution status */
  status: "success" | "failure";
}

File: cvc-lsp/src/protocol.rs
use cvc_core::models::Author;
use cvc_core::vscode::ResponsePartRaw;
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct SessionStartParams {
    pub title: String,
    pub timestamp: Option<i64>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TurnStartParams {
    pub id: String, // Correlation ID for start/end
    pub prompt: String,
    pub author: Author, // Type-safe enum
    pub context_files: Option<Vec<String>>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TurnEndParams {
    pub id: String, // Correlation ID
    pub response: Option<String>,
    pub chain_of_thought: Option<String>,
    pub model: Option<String>,
    pub raw_response: Option<Vec<ResponsePartRaw>>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct LinkCommitParams {
    pub commit_sha: String,
    pub interaction_ids: Vec<String>,
}

// --- Batch Turn Types (for Chat Session Watcher) ---

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TurnBatchParams {
    /// The VS Code requestId - used for deduplication.
    /// When re-processing the same request, old interactions are deleted first.
    pub source_request_id: String,
    /// The session/conversation ID to group interactions under.
    pub session_id: String,
    /// Model name/identifier used for all segments.
    pub model: Option<String>,
    /// Ordered list of interaction segments (human turn first, then agent turns).
    pub interactions: Vec<InteractionSegment>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InteractionSegment {
    /// Author of this segment: "human" for the initial prompt turn, "agent" for model reasoning turns.
    pub author: Author,
    /// The user prompt (only present on the first/human segment).
    pub user_prompt: Option<String>,
    /// Chain of thought / thinking block (only on agent segments).
    pub chain_of_thought: Option<String>,
    /// The visible response text for this segment.
    pub response: Option<String>,
    /// Context files attached to this segment (typically only on human segment).
    pub context_files: Option<Vec<String>>,
}

// --- Timeline Request/Response Types ---

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TimelineGetParams {
    pub max_items: Option<u32>,
    pub include_unbound: Option<bool>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TimelineGetResponse {
    pub pending: Vec<InteractionSummary>,
    pub commits: Vec<CommitWithThoughts>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InteractionSummary {
    pub id: String,
    pub prompt_preview: String,
    pub timestamp: i64,
    pub author: String,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CommitWithThoughts {
    pub sha: String,
    pub message: String,
    pub timestamp: i64,
    pub thoughts: Vec<InteractionSummary>,
}

// --- Interaction Detail Request/Response Types ---

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InteractionGetParams {
    pub id: String,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InteractionDetail {
    pub id: String,
    pub timestamp: i64,
    pub author: String,
    pub user_prompt: String,
    pub model_name: Option<String>,
    pub model_response: Option<String>,
    pub chain_of_thought: Option<String>,
    pub linked_commit: Option<String>,
    pub context_files: Vec<ContextFileInfo>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ContextFileInfo {
    pub file_path: String,
    pub start_line: Option<i32>,
    pub end_line: Option<i32>,
}
